server/server/open_telemetry_util.py

```diff
@@ -0,0 +1,112 @@
+import os
+import sys
+
+import requests
+import structlog
+from opentelemetry import trace
+from opentelemetry.exporter.cloud_trace import CloudTraceSpanExporter
+from opentelemetry.instrumentation.django import DjangoInstrumentor
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.propagators.cloud_trace_propagator import (
+    CloudTraceFormatPropagator,
+)
+from opentelemetry.resourcedetector.gcp_resource_detector import (
+    GoogleCloudResourceDetector,
+)
+from opentelemetry.sdk.resources import ProcessResourceDetector
+from opentelemetry.sdk.trace import TracerProvider, sampling
+from opentelemetry.sdk.trace.export import (
+    ConsoleSpanExporter,
+    SimpleSpanProcessor,
+)
+
+from server.config_util import get_project_config, is_running_tests
+from server.logging_util import add_metadata_to_all_logs_for_current_request
+
+
+def instrument_app():
+    config = get_project_config()
+    IS_LOCAL_DEV = config("IS_LOCAL_DEV", cast=bool, default=False)
+    DEV_TELEMETRY_CONSOLE_OUTPUT = config(
+        "DEV_TELEMETRY_CONSOLE_OUTPUT", cast=bool, default=False
+    )
+
+    if IS_LOCAL_DEV:
+        project_id = "local-dev"
+
+        span_exporter = ConsoleSpanExporter(
+            out=(
+                sys.stdout
+                if DEV_TELEMETRY_CONSOLE_OUTPUT and not is_running_tests()
+                else open(os.devnull, "w")
+            )
+        )
+
+        resource = ProcessResourceDetector(raise_on_error=True).detect()
+    else:
+        project_id = requests.get(
+            "http://metadata.google.internal/computeMetadata/v1/project/project-id",
+            headers={"Metadata-Flavor": "Google"},
+        ).text
```

Comment:
We have to make an HTTP request just to figure out the project ID? That just seems really strange, isn't it usually also available as an environment variable?

Reply:
Not in Cloud Run, by default. Agree that it's weird, I know I had some reason in mind for going with this, maybe just because it's the "official" way to do things? Could look for an env var first and fall back on this.